datafusion/optimizer/src/analyzer/mod.rs

```diff
@@ -140,6 +143,14 @@ impl Analyzer {
     where
         F: FnMut(&LogicalPlan, &dyn AnalyzerRule),
     {
+        // verify the logical plan required invariants at the start, before analyzer
+        plan.check_invariants(InvariantLevel::Always).map_err(|e| {
+            DataFusionError::Context(
+                "assert_lp_invariants_before_analyzers".to_string(),
```

Comment:
I also think a more human readable message might be nice, for example instead of 

"assert_lp_invariants_before_analyzers"

Maybe something like 

"Invalid input plan passed to Analyzer"